File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # preorder 실제로 시행하는 recursive 함수 정의
    # visit 표시하기 위한 ans 를 반환
    def traverse(self, ans: List[int], root: Optional[TreeNode]):
        if root is not None:
            ans.append(root.val)
            ans = self.traverse(ans, root.left)
            ans = self.traverse(ans, root.right)
        return ans

    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        ans = self.traverse(ans, root)
        return ans

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

// 함수에서 반환값 2개로 하기 위한 구조체 정의
typedef struct ret {
    int *ans;
    int cnt;
} ret;

// preorder 실제로 시행하는 recursive 함수 정의
// visit 표시하기 위한 ans, ans의 원소 개수를 표시하기 위한 cnt
ret traverse(int* ans, struct TreeNode* root, int cnt) {
    ret returnValue;
    returnValue.ans = ans;
    returnValue.cnt = cnt;

    if (root != 0) {
        returnValue.ans[returnValue.cnt] = root->val;
        returnValue.cnt += 1;

        returnValue = traverse(returnValue.ans, root->left, returnValue.cnt);
        returnValue = traverse(returnValue.ans, root->right, returnValue.cnt);
    }
    return returnValue;
}

int* preorderTraversal(struct TreeNode* root, int* returnSize) {
    int *ans = (int *)malloc(sizeof(int) * 100);
    int cnt = 0;
    
    ret returnValue;
    returnValue = traverse(ans, root, cnt);
    
    *returnSize = returnValue.cnt;
    return returnValue.ans;
}